[enhancement]shasum error. Because is not compatible with cnpmjs.

## problem

It occurs "shasum check failed" error.
the whole sentence is 

```
npm ERR! shasum check failed for /var/folders/cc/n1t9n30s45509jsbws5v723h0000gn/T/npm-61661-e421db3b/192.168.174.233_5080/tarballs/koa-static/1.5.2.tgz
```

and the `cat ....koa-static/1.5.2.tgz` is

```
{"error":"not_found","reason":"document not found"}%
```
## why

This error comes by setting a [cnpmjs](https://github.com/cnpm/cnpmjs.org) repository as a "--remote" fullfatdb. 
It may be because of the different strategy of the url for downloading.
## fix

I think this is a lack function more than a bug.
And have no idea because have not look into the code.
But a fix for this will greatly helpful. (https://registry.npmjs.org here is slow)

I think that may be my wrong expectation about how to use local-npm with --remote and --remote-skim.

A fast npm registry mirror here in China is https://registry.npm.taobao.org/ , which is built by [cnpmjs](https://cnpmjs.org/).
And It provide no skim-db, so I try to run local-npm by 
```
local-npm --url-base http://192.168.0.48:5080 --remote https://registry.npm.taobao.org/
```

I run `npm install 1grids` with this registry and it return errors:
```
npm ERR! Darwin 16.4.0
npm ERR! argv "/usr/local/Cellar/node/7.4.0/bin/node" "/usr/local/bin/npm" "install" "1grids"
npm ERR! node v7.4.0
npm ERR! npm  v4.0.5

npm ERR! shasum check failed for /var/folders/bl/749wcppx5c1724jwyd00t8xh0000gq/T/npm-25864-825d51d8/192.168.0.48_5080/tarballs/1grids/4.0.6.tgz
npm ERR! Expected: d01bb78a00d88fcc1f9100042b98dd2671111abe
npm ERR! Actual:   8d065612f683d9f8df207e665ab24398309c912e
npm ERR! From:     http://192.168.0.48:5080/tarballs/1grids/4.0.6.tgz
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     /private/tmp/aaaa/npm-debug.log
```

while 
```
>curl http://192.168.0.48:5080/tarballs/1grids/4.0.6.tgz
{"error":"not_found","reason":"document not found"}%
```

---------------------

I try `npm show 1grids` using https://registry.npm.taobao.org/, and I get tarball:
```
dist:
   { shasum: 'd01bb78a00d88fcc1f9100042b98dd2671111abe',
     size: 4639,
     noattachment: false,
     tarball: 'http://registry.npm.taobao.org/1grids/download/1grids-4.0.6.tgz' }
```
other than using https://registry.npmjs.org
```
dist:
   { shasum: 'd01bb78a00d88fcc1f9100042b98dd2671111abe',
     tarball: 'https://registry.npmjs.org/1grids/-/1grids-4.0.6.tgz' }
```

I think this could be a reason why comes error.

----------------------------------

This is not so boring me, because using registry.npmjs.org seems acceptable.

--- a/.eslintrc
+++ b/.eslintrc
@@ -1,25 +1,15 @@
 {
-  "parser": "babel-eslint",
-  "rules": {
-    "strict": 0,
-    "semi": [
-      2,
-      "always"
-    ],
-    "indent": [
-      2,
-      2,
-      {
-        "SwitchCase": 1
-      }
-    ],
-    "no-unused-vars": [
-      2,
-      {
-        "vars": "all",
-        "args": "after-used"
-      }
-    ],
-    "global-require": 2
-  }
+    "extends": "eslint:recommended",
+    "env": {
+        "node": true,
+        "mocha": true,
+        "es6": true
+    },
+    "rules": {
+        "no-constant-condition": 0
+    },
+    "parserOptions": {
+        "ecmaVersion": 2017,
+        "sourceType": "module"
+    }
 }

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,3 +4,4 @@
 - deprecates the support for `node@0.12`
 - fixes issue with not forwarding non-install commands
 - updates dependencies
+- fixes assumption that tarball location is predefined. 

--- a/lib/index.js
+++ b/lib/index.js
@@ -105,8 +105,8 @@ module.exports = function (argv) {
         // on-the-fly
         delete doc.versions[version];
       } else {
-        var tgz = urlBase + '/' + 'tarballs/' + name + '/' + version + '.tgz';
-        doc.versions[version].dist.tarball = tgz;
+        doc.versions[version].dist.tarball = urlBase + '/' + 'tarballs/' + name + '/' + version + '.tgz';
+        doc.versions[version].dist.info = urlBase + '/' + name + '/' + version;
       }
     }
     return doc;
@@ -183,7 +183,6 @@ module.exports = function (argv) {
     });
   });
   app.get('/tarballs/:name/:version.tgz', function (req, res) {
-
     var pkgName = req.params.name;
     var pkgVersion = req.params.version;
     var id = pkgName + '-' + pkgVersion;
@@ -198,12 +197,13 @@ module.exports = function (argv) {
       logger.timeEnd('2: skimLocal.get(' + pkgName + ')');
       // if we're daisy-chaining multiple local-npm's together,
       // then the remote tarball URL has to be swapped on-the-fly
+      var dist = doc.versions[pkgVersion].dist;
       if (fatRemoteIsAnotherLocalNpm) {
         doc = massageMetadata(FAT_REMOTE, doc);
       }
       var dist = doc.versions[pkgVersion].dist;
       logger.time('db.get(' + id + ')');
-      db.get(id, {asBuffer: true, valueEncoding: 'binary'}, function (err, buffer) {
+      db.get(id, { asBuffer: true, valueEncoding: 'binary' }, function (err, buffer) {
         logger.timeEnd('db.get(' + id + ')');
         if (!err) {
           var hash = crypto.createHash('sha1');
@@ -217,26 +217,64 @@ module.exports = function (argv) {
           }
         }
         logger.miss(pkgName, pkgVersion);
-        var buffs = [];
-        logger.time('request.get(' + dist.tarball + ')');
-        request.get(dist.tarball).on('error', function (err) {
-          logger.info(err);
-          res.status(500).send('you are offline and this package isn\'t cached');
-        }).pipe(through(function (chunk, _, next) {
-          buffs.push(chunk);
-          next();
-        }, function (next) {
-          next();
-          var buffer = Buffer.concat(buffs);
-          logger.timeEnd('request.get(' + dist.tarball + ')');
-          sendBinary(res, buffer);
-          db.put(id, buffer, function (err) {
-            logger.cached(pkgName, pkgVersion);
-            if (err) {
+
+        if(dist.info) {
+            logger.time('request.get(' + dist.info + ')');
+            var infoBuffs = [];
+            request.get(dist.info).on('error', function (err) {
               logger.info(err);
-            }
-          });
-        }));
+              res.status(500).send('you are offline and this package isn\'t cached');
+                }).pipe(through(function (chunk, _, next) {
+              infoBuffs.push(chunk);
+              next();
+            }, function (next) {
+              next();
+              var info = JSON.parse(Buffer.concat(infoBuffs));
+
+              var buffs = [];
+              request.get(info.dist.tarball).on('error', function (err) {
+                logger.info(err);
+                res.status(500).send('you are offline and this package isn\'t cached');
+              }).pipe(through(function (chunk, _, next) {
+                buffs.push(chunk);
+                next();
+              }, function (next) {
+                next();
+                var buffer = Buffer.concat(buffs);
+
+                logger.timeEnd('request.get(' + dist.info + ')');
+                sendBinary(res, buffer);
+                db.put(id, buffer, function (err) {
+                  logger.cached(pkgName, pkgVersion);
+                  if (err) {
+                    logger.info(err);
+                  }
+                });
+              }));
+            }));
+        } else {
+            logger.time('request.get(' + dist.tarball + ')');
+            var buffs = [];
+            request.get(dist.tarball).on('error', function (err) {
+              logger.info(err);
+              res.status(500).send('you are offline and this package isn\'t cached');
+            }).pipe(through(function (chunk, _, next) {
+              buffs.push(chunk);
+              next();
+            }, function (next) {
+              next();
+              var buffer = Buffer.concat(buffs);
+
+              logger.timeEnd('request.get(' + dist.tarball + ')');
+              sendBinary(res, buffer);
+              db.put(id, buffer, function (err) {
+                logger.cached(pkgName, pkgVersion);
+                if (err) {
+                  logger.info(err);
+                }
+              });
+            }));
+        }
       });
     }).catch(function (err) {
       logger.info(err);

--- a/lib/index.js
+++ b/lib/index.js
@@ -125,14 +125,6 @@ module.exports = function (argv) {
     res.set('Cache-Control', 'max-age=300');
   }
 
-  // TODO: this is an error-prone way to check this. We should probably
-  // send an HTTP request to it and check the response to detect local-npm.
-  var fatRemoteIsAnotherLocalNpm =
-    FAT_REMOTE !== 'https://registry.npmjs.org' &&
-    FAT_REMOTE !== 'https://registry.npmjs.org/' &&
-    FAT_REMOTE !== 'http://registry.npmjs.org' &&
-    FAT_REMOTE !== 'http://registry.npmjs.org/';
-
   //
   // actual server logic
   //
@@ -195,11 +187,6 @@ module.exports = function (argv) {
       throw new Error('offline');
     }).then(function (doc) {
       logger.timeEnd('2: skimLocal.get(' + pkgName + ')');
-      // if we're daisy-chaining multiple local-npm's together,
-      // then the remote tarball URL has to be swapped on-the-fly
-      if (fatRemoteIsAnotherLocalNpm) {
-        doc = massageMetadata(FAT_REMOTE, doc);
-      }
       var dist = doc.versions[pkgVersion].dist;
       logger.time('db.get(' + id + ')');
       db.get(id, { asBuffer: true, valueEncoding: 'binary' }, function (err, buffer) {

--- a/lib/index.js
+++ b/lib/index.js
@@ -197,7 +197,6 @@ module.exports = function (argv) {
       logger.timeEnd('2: skimLocal.get(' + pkgName + ')');
       // if we're daisy-chaining multiple local-npm's together,
       // then the remote tarball URL has to be swapped on-the-fly
-      var dist = doc.versions[pkgVersion].dist;
       if (fatRemoteIsAnotherLocalNpm) {
         doc = massageMetadata(FAT_REMOTE, doc);
       }
@@ -222,35 +221,35 @@ module.exports = function (argv) {
             logger.time('request.get(' + dist.info + ')');
             var infoBuffs = [];
             request.get(dist.info).on('error', function (err) {
-              logger.info(err);
-              res.status(500).send('you are offline and this package isn\'t cached');
-                }).pipe(through(function (chunk, _, next) {
-              infoBuffs.push(chunk);
-              next();
-            }, function (next) {
-              next();
-              var info = JSON.parse(Buffer.concat(infoBuffs));
-
-              var buffs = [];
-              request.get(info.dist.tarball).on('error', function (err) {
                 logger.info(err);
                 res.status(500).send('you are offline and this package isn\'t cached');
-              }).pipe(through(function (chunk, _, next) {
-                buffs.push(chunk);
+            }).pipe(through(function (chunk, _, next) {
+                infoBuffs.push(chunk);
                 next();
-              }, function (next) {
+            }, function (next) {
                 next();
-                var buffer = Buffer.concat(buffs);
+                var info = JSON.parse(Buffer.concat(infoBuffs));
 
-                logger.timeEnd('request.get(' + dist.info + ')');
-                sendBinary(res, buffer);
-                db.put(id, buffer, function (err) {
-                  logger.cached(pkgName, pkgVersion);
-                  if (err) {
+                var buffs = [];
+                request.get(info.dist.tarball).on('error', function (err) {
                     logger.info(err);
-                  }
-                });
-              }));
+                    res.status(500).send('you are offline and this package isn\'t cached');
+                }).pipe(through(function (chunk, _, next) {
+                    buffs.push(chunk);
+                    next();
+                }, function (next) {
+                    next();
+                    var buffer = Buffer.concat(buffs);
+
+                    logger.timeEnd('request.get(' + dist.info + ')');
+                    sendBinary(res, buffer);
+                    db.put(id, buffer, function (err) {
+                        logger.cached(pkgName, pkgVersion);
+                        if (err) {
+                            logger.info(err);
+                        }
+                    });
+                }));
             }));
         } else {
             logger.time('request.get(' + dist.tarball + ')');

--- a/lib/logger.js
+++ b/lib/logger.js
@@ -5,7 +5,7 @@ var argv = require('yargs').argv;
 var level  = require('./levels')(argv.l);
 
 function log(msg) {
-  console.log(msg);
+  console.log(msg); // eslint-disable-line
 }
 
 function wrap(thing) {
@@ -25,13 +25,13 @@ function wrap(thing) {
 exports.time = function (msg) {
   msg = wrap(msg);
   if (level > 2) {
-    console.time(msg);
+    console.time(msg); // eslint-disable-line
   }
 };
 exports.timeEnd = function (msg) {
   msg = wrap(msg);
   if (level > 2) {
-    console.timeEnd(msg);
+    console.timeEnd(msg); // eslint-disable-line
   }
 };
 exports.info = function (msg) {
@@ -85,6 +85,6 @@ exports.cached = function (pkg, version) {
 exports.warn = function (msg) {
   msg = wrap(msg);
   if (level > 0) {
-    log(msg.yellow); 
+    log(msg.yellow);
   }
 };

--- a/package.json
+++ b/package.json
@@ -47,7 +47,6 @@
   },
   "devDependencies": {
     "babel-cli": "^6.4.0",
-    "babel-eslint": "^7.2.3",
     "babel-plugin-syntax-async-functions": "^6.3.13",
     "babel-plugin-transform-regenerator": "^6.3.26",
     "babel-preset-es2015": "^6.3.13",

--- a/package.json
+++ b/package.json
@@ -58,6 +58,7 @@
     "istanbul": "^0.4.1",
     "istanbul-coveralls": "^1.0.3",
     "memdown": "^1.2.4",
+    "mkdirp": "^0.5.1",
     "mocha": "^3.4.1",
     "ncp": "^2.0.0",
     "nock": "^9.0.13",

--- a/package.json
+++ b/package.json
@@ -58,7 +58,6 @@
     "istanbul": "^0.4.1",
     "istanbul-coveralls": "^1.0.3",
     "memdown": "^1.2.4",
-    "mkdirp": "^0.5.1",
     "mocha": "^3.4.1",
     "ncp": "^2.0.0",
     "nock": "^9.0.13",

--- a/test/test-offline.js
+++ b/test/test-offline.js
@@ -9,13 +9,13 @@
 
 var nock = require('nock');
 
-console.log('mocking offline state with nock...');
+console.log('mocking offline state with nock...'); // eslint-disable-line
 nock('https://skimdb.npmjs.com:443', {"encodedQueryParams":true})
   .get('/registry/')
   .replyWithError('oh no you are offline!');
 
 setTimeout(function () {
-  console.log('no more mocking');
+  console.log('no more mocking'); // eslint-disable-line
   nock.restore();
 }, 5000);
 

--- a/test/test.js
+++ b/test/test.js
@@ -287,8 +287,6 @@ describe('main test suite', function () {
       .map(row => row.id)
       .filter(id => blacklist.indexOf(id) == -1);
 
-    console.log(packages);
-
     await ncp('./test/project3', WORK_DIR);
 
     // check that they are npm installed

